File: output/BMSDiagnosticServiceKotlin/BMSDiagnosticServiceKotlin.java
import com.visteon.someip.SomeIP;
import com.visteon.someip.Service;
import com.visteon.someip.Method;
import com.visteon.someip.Event;
import com.visteon.someip.Message;
import com.visteon.someip.Payload;
import com.visteon.someip.StatusCode;

public class BMSDiagnosticServiceKotlin {
    private Service service;
    private Event batteryWarningEvent;

    public BMSDiagnosticServiceKotlin() {
        this.service = new Service(4097, 1);
        this.batteryWarningEvent = new Event(32769, 1);

        Method getBatteryStatusMethod = new Method(1, () -> {
            float[] batteryStatus = getBatteryStatus();
            Payload payload = new Payload(new Object[]{batteryStatus[0], batteryStatus[1], batteryStatus[2], batteryStatus[3], batteryStatus[4]});
            return new Message(payload);
        });

        Method getCellVoltagesMethod = new Method(2, () -> {
            float[] cellVoltages = getCellVoltages();
            Payload payload = new Payload(new Object[]{cellVoltages});
            return new Message(payload);
        });

        Method getEstimatedRangeMethod = new Method(3, (params) -> {
            int drivingMode = (int) params[0];
            float rangeKm = getEstimatedRange(drivingMode);
            Payload payload = new Payload(new Object[]{rangeKm});
            return new Message(payload);
        });

        this.service.addMethod(getBatteryStatusMethod);
        this.service.addMethod(getCellVoltagesMethod);
        this.service.addMethod(getEstimatedRangeMethod);
        this.service.addEvent(batteryWarningEvent);

        SomeIP.getInstance().registerService(this.service);
    }

    private float[] getBatteryStatus() {
        // Implement logic to retrieve battery status
        return new float[]{25.0f, 12.6f, -3.4f, 28.0f, 1};
    }

    private float[] getCellVoltages() {
        // Implement logic to retrieve cell voltages
        return new float[]{3.7f, 3.8f, 3.9f};
    }

    private float getEstimatedRange(int drivingMode) {
        // Implement logic to estimate range based on driving mode
        return 150.0f;
    }

    public void checkBatteryStatus() {
        float[] batteryStatus = getBatteryStatus();
        if (batteryStatus[0] < 20) {
            emitWarning(0x0001, "Low battery");
        }
        if (batteryStatus[3] > 45) {
            emitWarning(0x0002, "High temperature");
        }
        if (batteryStatus[3] > 60) {
            throw new RuntimeException("Critical temperature - shutdown required", new Exception());
        }
    }

    private void emitWarning(int code, String message) {
        Payload payload = new Payload(new Object[]{code, message});
        Message warningMessage = new Message(payload);
        batteryWarningEvent.emit(warningMessage);
    }

    public static void main(String[] args) {
        BMSDiagnosticServiceKotlin service = new BMSDiagnosticServiceKotlin();
        service.checkBatteryStatus();
    }
}

File: output/BMSDiagnosticServiceKotlin/bmsdiagnosticservicekotlin.kt
import java.util.logging.Logger

class BMSDiagnosticServiceKotlin {
    private val logger = Logger.getLogger(BMSDiagnosticServiceKotlin::class.java.name)
    private var batterySOC: Float = 0.0f
    private var batteryVoltage: Float = 0.0f
    private var batteryCurrent: Float = 0.0f
    private var batteryTemperature: Float = 0.0f

    fun GetBatteryStatus(): Map<String, Any> {
        val status = mapOf(
            "soc" to batterySOC,
            "voltage" to batteryVoltage,
            "current" to batteryCurrent,
            "temperature" to batteryTemperature,
            "health_status" to getHealthStatus()
        )

        checkBatteryStatus(status)
        return status
    }

    fun GetCellVoltages(): Map<String, Any> {
        // Implement logic to get cell voltages
        val cellVoltages = listOf(3.7f, 3.8f, 3.9f)
        return mapOf("cell_voltages" to cellVoltages)
    }

    fun GetEstimatedRange(drivingMode: Int): Map<String, Any> {
        // Implement logic to get estimated range based on driving mode
        val rangeKM = 200.0f * (1 - batterySOC / 100)
        return mapOf("range_km" to rangeKM)
    }

    private fun getHealthStatus(): Byte {
        if (batterySOC < 20) return 0x01
        if (batteryTemperature > 45) return 0x02
        if (batteryTemperature > 60) throw Exception("Critical temperature - shutdown required")
        return 0x00
    }

    private fun checkBatteryStatus(status: Map<String, Any>) {
        val soc = status["soc"] as Float
        val voltage = status["voltage"] as Float
        val current = status["current"] as Float
        val temperature = status["temperature"] as Float

        if (soc < 20) throw Exception("Low battery")
        if (temperature > 45) throw Exception("High temperature")
        if (temperature > 60) throw Exception("Critical temperature - shutdown required")

        logger.info("Battery status: SOC=$soc%, Voltage=$voltageV, Current=$currentA, Temperature=$temperatureC")
    }
}